core/account-id/src/errors.rs

```diff
@@ -38,24 +38,27 @@ pub enum ParseErrorKind {
     /// ends with or has separators immediately following each other.
     ///
     /// Cases: `jane.`, `angela__moss`, `tyrell..wellick`
-    RedundantSeparator,
+    RedundantSeparator(usize, char),
     /// The Account ID contains an invalid character.
     ///
     /// This variant would be returned if the Account ID contains an upper-case character, non-separating symbol or space.
     ///
     /// Cases: `ƒelicia.near`, `user@app.com`, `Emily.near`.
-    InvalidChar,
+    InvalidChar(usize, char),
```

Comment:
I would define these as inline structs rather tuples to be more explicit with the field names:

```
InvalidChar { invalid_char: char, position: usize }
```

P.S. let's also add a comment that position is the index of the Unicode character rather than an index in a byte array representing the account id string.

Reply:
>P.S. let's also add a comment that position is the index of the Unicode character rather than an index in a byte array representing the account id string.

Wouldn't the two be the same? We only allow ascii, and we'll bail of the first error. 

I would say that, in the Rust API, we *should* use utf8 indices, and let the caller to convert from that to grapheme clusters or whatever else is appropriate for display. Index of unicode character is generally a wrong thing to use for any purpose.

Reply:
> I would say that, in the Rust API, we should use utf8 indices, and let the caller to convert from that to grapheme clusters or whatever else is appropriate for display. Index of unicode character is generally a wrong thing to use for any purpose.

Fair enough

Reply:
> Index of unicode character is generally a wrong thing to use for any purpose

To be clear, are you suggesting we drop `char ([u8; 4])` and read by bytes instead, much like before?

That would be convenient considering we only accept ASCII anyway, but that impairs the API a bit if we have this PR merged.

`InvalidChar { pos: usize, char: char }` is a lot more helpful than `InvalidChar { pos: usize, char: u8 }`

It's also not very clear what you meant by "utf8 indices".. There's `.char_indices()` for positional consistency, there's `.chars().enumerate()` for char size indifference, and there's raw bytes.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d73bf43f591e42855c9b5783225f6cb1

Reply:
> I would define these as inline structs rather tuples to be more explicit with the field names:

Yeah, although I was hoping to model that after, for instance – `.enumerate()` with `(usize, item)`.

Plus, considering it's just two fields, it's not that hard to figure out. But I'll push a patch on another PR, so we can take a look and compare.

Reply:
\+ looking at semver, for instance, indexing is byte-wise but encodes the character at the invalid byte position as an actual `char` into the `ErrorKind`.

https://github.com/dtolnay/semver/blob/98163cf2fc/src/parse.rs#L177

Reply:
`semver` is a very good example, thanks for bringing it up! It doesn't expose error kind at all: https://docs.rs/semver/latest/semver/struct.Error.html. This seems like a nice simplification. Do we have a concrete use-case for the consumer to inspect the kind?

Reply:
>To be clear, are you suggesting we drop char ([u8; 4]) and read by bytes instead, much like before?

My original suggestion was to keep the impl as is, but document that `usize` is a byte index. That is, that, if you get a `pos` in the error when parsing `text` as AccountId, `&text[pos..]` is a valid `&str`. But now I lean towards not exposing the Kind at all.

Reply:
> Do we have a concrete use-case for the consumer to inspect the kind?

Not really, but I can imagine interfaces providing useful hints on Account ID construction.

I mean, for one, much like with `semver`: `serde` + `borsh` now helpfully report the position at which the ID was flagged to be invalid. Although it wouldn't hurt to retain that provision privately.

It's already beneficial that we have that information when `Display`-ed, it's unlikely users would need to be matching error kinds as well as position hints. So I'll drop it from the public API, but retain the hints when `Display`-ed.